qa: Fix name collisions

kotreshhr opened the pull request:
Addresses the name collisions of volumes, subvolumes,
clones, subvolume groups and snapshots within the tests.

Fixes: https://tracker.ceph.com/issues/43517
Signed-off-by: Kotresh HR <khiremat@redhat.com>


<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

kotreshhr commented:
jenkins test dashboard backend

kotreshhr commented:
> @kotreshhr need a bit more info about the issue. Subvolume (and other) names were are randomized for each tests.

The issue was, with in the tests, if multiple subvolumes were used, those names were colliding when using generate random names functions. So I have removed all those instances.

kotreshhr commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -639,7 +639,7 @@ def test_subvolumegroup_pin_distributed(self):
         self._fs_cmd("subvolumegroup", "pin", self.volname, group, "export", "0")
         for i in range(10):
-            subvolume = self._generate_random_subvolume_name()
+            subvolume = "subvolume_{0}".format(i)
```
My bad, I misread the document as the returned list might not be unique but it is indeed unique. I will refresh the patch retaining these instances.

kotreshhr replied:
But this specific instance before might cause collision as random names are generated in the sample of 100 each time in the loop.

vshankar replied:
right -- that the reason to use random.sample()

kotreshhr replied:
Done

vshankar commented:
> The issue was, with in the tests, if multiple subvolumes were used, those names were colliding when using generate random names functions. So I have removed all those instances.

I had seen this earlier and fixed it in commit b9cff8af8e8f423c8e2b58de2b4da48842223cf3

kotreshhr commented:
> > The issue was, with in the tests, if multiple subvolumes were used, those names were colliding when using generate random names functions. So I have removed all those instances.
> 
> I had seen this earlier and fixed it in commit [b9cff8a](https://github.com/ceph/ceph/commit/b9cff8af8e8f423c8e2b58de2b4da48842223cf3)

Yes, same issue but few more instances of incorrect usage of the function 'self._generate_random_subvolume_name' has crept in.

kotreshhr commented:
> Looks good, Kotresh.
> 
> Instead of generating random subvolume and other entities' names, can't we just generate them deterministically? e.g., self._generate_subvolume_name(4) will always give you "subvol_{0,1,2,3}"

We could do that. Since the the routine is used across the test suite and I didn't want to remove it. If we are not removing it, I thought using it correctly makes sense.

vshankar commented:
> Instead of generating random subvolume and other entities' names, can't we just generate them deterministically? e.g., self._generate_subvolume_name(4) will always give you "subvol_{0,1,2,3}"

I choose random subvolume names so as to quickly fetch the commands executed for the failed test in teuthology log.

ajarr commented:
> > Instead of generating random subvolume and other entities' names, can't we just generate them deterministically? e.g., self._generate_subvolume_name(4) will always give you "subvol_{0,1,2,3}"
> 
> I choose random subvolume names so as to quickly fetch the commands executed for the failed test in teuthology log.

I don't follow. Generating incremental numbers as suffixes is simple and quick.

vshankar commented:
> I don't follow. Generating incremental numbers as suffixes is simple and quick.

incremental for each test -- that's fine. all this is more of my specificity w/ debugging rather than anything to do with functionality.

ajarr commented:
> > Looks good, Kotresh.
> > Instead of generating random subvolume and other entities' names, can't we just generate them deterministically? e.g., self._generate_subvolume_name(4) will always give you "subvol_{0,1,2,3}"
> 
> We could do that. Since the the routine is used across the test suite and I didn't want to remove it. If we are not removing it, I thought using it correctly makes sense.

Discussed with Venky, looks like he opted for random numbers/entity names to help him debug test failures quicker.  The tracker ticket description says,
```
I'm not really sure why we need to produce random groups here to begin with; should they not just be arbitrary static names?
```
@batrick, do you want the PR to switch to using static names?

batrick commented:
How about: generate a random number for the start of a deterministic sequence for each test. So instead:

````
def setUp(self):
  self.subvolume_start = random.randint(1, (1<<20))
  ...

def generate_subvolume_names(self, count):
        n = self.subvolume_start
        ret = [f"{TestVolumes.TEST_VOLUME_PREFIX}_{i:016}" for i in range(n, n+count)]
        self.subvolume_start += count
        return ret
````

kotreshhr commented:
I am fine with this approach to avoid the coding errors that has led to this issue.

> How about: generate a random number for the start of a deterministic sequence for each test. So instead:
> 
> ```
> def setUp(self):
>   self.subvolume_start = random.randint(1, (1<<20))

   Do we really need to start with random number here ? Why can't this be initialized to 1? It should work with no collisions as we delete volume between each tests right?

>   ...
> 
> def generate_subvolume_names(self, count):
>         n = self.subvolume_start
>         ret = [f"{TestVolumes.TEST_VOLUME_PREFIX}_{i:016}" for i in range(n, n+count)]
>         self.subvolume_start += count
>         return ret
> ```

ajarr commented:
> I am fine with this approach to avoid the coding errors that has led to this issue.
> 
> > How about: generate a random number for the start of a deterministic sequence for each test. So instead:
> > ```
> > def setUp(self):
> >   self.subvolume_start = random.randint(1, (1<<20))
> > ```
> 
> Do we really need to start with random number here ? Why can't this be initialized to 1? It should work with no collisions as we delete volume between each tests right?
> 

Venky generated random number based names to help him debug tests quicker. I think the rationale was that  when going through the ceph daemon logs, information about a volume entity e.g., a subvolume,  with a "unique" name (random number suffix) could all be tracked to a single test function.

The solution of generating a random initial number for a test and then using sequential numbers for the test solves the subvolume collision issue and possibly helps in debugging test failures.

> > ...
> > def generate_subvolume_names(self, count):
> > n = self.subvolume_start
> > ret = [f"{TestVolumes.TEST_VOLUME_PREFIX}_{i:016}" for i in range(n, n+count)]
> > self.subvolume_start += count
> > return ret

vshankar commented:
> How about: generate a random number for the start of a deterministic sequence for each test. So instead:
> 
> ```
> def setUp(self):
>   self.subvolume_start = random.randint(1, (1<<20))
>   ...
> 
> def generate_subvolume_names(self, count):
>         n = self.subvolume_start
>         ret = [f"{TestVolumes.TEST_VOLUME_PREFIX}_{i:016}" for i in range(n, n+count)]
>         self.subvolume_start += count
>         return ret
> ```

nice and clean approach

kotreshhr commented:
> 
> Venky generated random number based names to help him debug tests quicker. I think the rationale was that when going through the ceph daemon logs, information about a volume entity e.g., a subvolume, with a "unique" name (random number suffix) could all be tracked to a single test function.
> 
> The solution of generating a random initial number for a test and then using sequential numbers for the test solves the subvolume collision issue and possibly helps in debugging test failures.
> 

Thanks for the explanation Ramana! It helped. I will go with Patrick's approach and refresh the PR.

kotreshhr commented:
jenkins test dashboard backend

kotreshhr commented:
jenkins test dashboard backend

batrick commented:
jenkins retest this please

batrick commented:
https://pulpito.ceph.com/?branch=wip-pdonnell-testing-20200717.003135

Unrelated new failures:

- https://tracker.ceph.com/issues/46609
- https://tracker.ceph.com/issues/46608
